### PR TITLE
View notifications by repository

### LINKED ISSUE #61: Support for filtering/grouping of notifications by repository
Freetime currently displays a list of notifications from a user's public repositories (notifications from private repositories are currently not displayed, see #60 for a discussion). These notifications are ordered via their timestamp in chronological order. The list may contain several hundred read/unread notifications. 

It will be great if we could apply a "filter" to select one or more repositories the user is subscribed for notifications and show notifications from only the selected repositories. In addition, notifications can possibly be grouped by repositories 🤔 . This will probably make things complicated as we would want pagination of notifications. 

image

### PR DESCRIPTION
**Pull request by rnystrom**
- New left/right nav bar items for inbox
- Show alert, select repo
- `NotificationsViewController` now handles unread, all, and repo (show all for repo)
- "Mark Read" for repo clears _only_ that repo's notifications

Pretty happy w/ this quick hack. Been using it all night. This will get us some immediate feedback before investing much in custom UI controls (e.g. the whole nav-bar dropdown thing).

Fixes #61 

img_0554

img_0555

### CODE REVIEW

**Comment by BasThomas** [comment]
What is the expected behavior for showing repos? If I have no unreads, does it automatically open archives? (and why has that button been renamed to "View Read"?)

File: Classes/Notifications/NotificationClient.swift [hunk 1]
```diff
@@ -71,16 +79,12 @@ final class NotificationClient {
...
-        githubClient.request(GithubClient.Request(
-            path: "notifications",
-            method: .put) { (response, _) in
-                guard let completion = completion else { return }
-                // https://developer.github.com/v3/activity/notifications/#mark-as-read
-                let success = response.response?.statusCode == 205
-                completion(success)
-        })
+    func markAllNotifications(completion: @escaping (Bool) -> Void) {
+        markNotifications(repo: nil, completion: completion)
```
Comment (BasThomas) [inline Classes/Notifications/NotificationClient.swift]: You can omit the `repo` parameter as it defaults to `nil`. :)

File: Classes/Notifications/NotificationClient.swift [hunk 2]
```diff
@@ -123,4 +127,49 @@ final class NotificationClient {
...
+                completion(.success(repos.sorted { $0.name < $1.name }))
+            } else {
+                completion(.error(response.error))
+            }
+        })
+    }
+
+    // MARK: Private API
+
+    func markNotifications(repo: (owner: String, name: String)?, completion: @escaping (Bool) -> Void) {
```
Comment (BasThomas) [inline Classes/Notifications/NotificationClient.swift]: Looks pretty similar to https://github.com/rnystrom/GitHawk/pull/1291/files#diff-d08adcb8b7b2ae5fc2e0f5abfaa92a42R34, maybe you can share some logic?

File: Classes/Notifications/NotificationClient.swift [hunk 3]
```diff
@@ -123,4 +127,49 @@ final class NotificationClient {
...
+            path = "repos/\(repo.owner)/\(repo.name)/notifications"
+        } else {
+            // https://developer.github.com/v3/activity/notifications/#mark-as-read
+            path = "notifications"
+        }
+
+        githubClient.request(GithubClient.Request(
+            path: path,
+            method: .put
+        ) { (response, _) in
```
Comment (BasThomas) [inline Classes/Notifications/NotificationClient.swift]: Nit, but you can omit the `()` for `(response, _)`.

File: Classes/Notifications/NotificationSubscriptionsController.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+//
+//  Created by Ryan Nystrom on 12/18/17.
+//  Copyright © 2017 Ryan Nystrom. All rights reserved.
+//
+
+import UIKit
+
+final class NotificationSubscriptionsController {
+
+    enum State {
```
Comment (BasThomas) [inline Classes/Notifications/NotificationSubscriptionsController.swift]: Where did you get this from? I love it 😀

File: Classes/Notifications/NotificationSubscriptionsController.swift [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                ToastManager.showGenericError()
+            }
+        }
+    }
+
+    // MARK: Private API
+
+    var refreshAction: UIAlertAction {
+        return UIAlertAction(
+            title: NSLocalizedString("Refresh", comment: ""),
```
Comment (BasThomas) [inline Classes/Notifications/NotificationSubscriptionsController.swift]: Wondering where this is used - why would we "hide" a refresh in a `UIAlertAction` if anywhere else we use a `UIRefreshControl`?

File: Classes/Notifications/NotificationSubscriptionsController.swift [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    func repoActions(repos: [Repository]) -> [UIAlertAction] {
+        return repos.map {
+            let repo = $0
+            return UIAlertAction(title: repo.name, style: .default, handler: { [weak self] _ in
+                self?.pushRepoNotifications(owner: repo.owner.login, name: repo.name)
+            })
+        }
+    }
+
+    func pushRepoNotifications(owner: String, name: String) {
```
Comment (BasThomas) [inline Classes/Notifications/NotificationSubscriptionsController.swift]: Maybe `pushRepoNotificationsController` or similar? Was a tiny bit confused with `git push` etc etc.

Comment (BasThomas) [reply Classes/Notifications/NotificationSubscriptionsController.swift]: Also (unrelated to this func), we are using a combination of `owner` and `name` quite a bit (especially in this PR). Maybe wrap it in a quick struct (or see if we can use an existing one; I bet we can)?

File: Classes/Notifications/NotificationsViewController.swift [hunk 7]
```diff
@@ -20,24 +20,41 @@ TabNavRootViewControllerType,
...
+        self.inboxType = inboxType
 
         super.init(
             emptyErrorMessage: NSLocalizedString("Cannot load your inbox.", comment: ""),
             dataSource: self
         )
 
         self.foreground.delegate = self
+
+        switch inboxType {
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: What about doing this logic in an `extension NotificationsViewController.InboxType` to prevent cluttering init?

File: Classes/Notifications/NotificationsViewController.swift [hunk 8]
```diff
@@ -47,53 +64,76 @@ FlatCacheListener {
     override func viewDidLoad() {
         super.viewDidLoad()
 
-        resetLeftBarItem()
-        if !showRead {
-            navigationItem.rightBarButtonItem = UIBarButtonItem(
-                title: NSLocalizedString("Archives", comment: ""),
+        resetRightBarItem()
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: Whyyyyy are you swapping the buttons?

File: Classes/Notifications/NotificationsViewController.swift [hunk 9]
```diff
@@ -47,53 +64,76 @@ FlatCacheListener {
...
+        switch inboxType {
+        case .unread:
+            let item = UIBarButtonItem(
+                image: UIImage(named: "bullets-hollow"),
                 style: .plain,
                 target: self,
-                action: #selector(NotificationsViewController.onViewAll)
+                action: #selector(NotificationsViewController.onMore)
             )
+            item.accessibilityLabel = NSLocalizedString("More options", comment: "")
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: 👊😍

File: Classes/Notifications/NotificationsViewController.swift [hunk 10]
```diff
@@ -47,53 +64,76 @@ FlatCacheListener {
     override func viewDidLoad() {
         super.viewDidLoad()
 
-        resetLeftBarItem()
-        if !showRead {
-            navigationItem.rightBarButtonItem = UIBarButtonItem(
-                title: NSLocalizedString("Archives", comment: ""),
+        resetRightBarItem()
+
+        switch inboxType {
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: `if case .unread = inboxType`? See my beautiful [blog post](https://basthomas.github.io/switches-and-cases).

File: Classes/Notifications/NotificationsViewController.swift [hunk 11]
```diff
@@ -47,53 +64,76 @@ FlatCacheListener {
...
     // MARK: Private API
 
-    @objc func onViewAll() {
-        let controller = NotificationsViewController(client: client, showRead: true)
-        controller.title = NSLocalizedString("Archives", comment: "")
+    @objc func onMore() {
+        let alert = UIAlertController.configured(preferredStyle: .actionSheet)
+
+        alert.add(action: UIAlertAction(
+            title: NSLocalizedString("View Read", comment: ""),
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: Also commented in the PR itself, but why did you rename this?

File: Classes/Notifications/NotificationsViewController.swift [hunk 12]
```diff
@@ -47,53 +64,76 @@ FlatCacheListener {
...
+            style: .default,
+            handler: { [weak self] _ in
+                self?.onViewAll()
+        }))
+        subscriptionController?.actions.forEach({ alert.add(action: $0) })
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: Can omit the `()`.

File: Classes/Notifications/NotificationsViewController.swift [hunk 13]
```diff
@@ -47,53 +64,76 @@ FlatCacheListener {
...
-            )
-        }
+    func resetRightBarItem() {
+        let item = UIBarButtonItem(
+            image: UIImage(named: "check"),
+            style: .plain,
+            target: self,
+            action: #selector(NotificationsViewController.onMarkAll)
+        )
+        item.accessibilityLabel = NSLocalizedString("", comment: "")
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: 🤔

File: Classes/Notifications/NotificationsViewController.swift [hunk 14]
```diff
@@ -108,18 +148,35 @@ FlatCacheListener {
...
     }
 
-    @objc private func onMarkAll(sender: UIBarButtonItem) {
+    @objc private func onMarkAll() {
+        let message: String
+        switch inboxType {
+        case .all, .unread:
+            message = NSLocalizedString("Mark all notifications as read?", comment: "")
+        case .repo(_, let name):
+            let messageFormat = NSLocalizedString("Mark %@ notifications as read?", comment: "")
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: Saw this and my initial thought was "shouldn't this be `%d` to show the number of notifications?" and then I saw that this is for the repo name. What about adding the amount of notifications to this message?

File: Classes/Notifications/NotificationsViewController.swift [hunk 15]
```diff
@@ -159,17 +216,37 @@ FlatCacheListener {
...
     override func fetch(page: NSNumber?) {
+        subscriptionController?.fetchSubscriptions()
+
         let width = view.bounds.width
+
+        let repo: (String, String)?
+        switch inboxType {
+        case .repo(let owner, let name):
+            repo = (owner, name)
+        case .all:
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: `case .all, .unread:`

File: Classes/Notifications/NotificationsViewController.swift [hunk 16]
```diff
@@ -184,9 +261,10 @@ FlatCacheListener {
     func models(listAdapter: ListAdapter) -> [ListDiffable] {
         var models = [NotificationViewModel]()
 
+        let showAll = self.showAll
```
Comment (BasThomas) [inline Classes/Notifications/NotificationsViewController.swift]: Why shadow?

File: Classes/Notifications/NotificationSubscriptionsController.swift (hunk 4 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationSubscriptionsController.swift]: Haha, what do you mean?

File: Classes/Notifications/NotificationSubscriptionsController.swift (hunk 5 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationSubscriptionsController.swift]: So you should almost never see this. I'm making sure that if the fetch request to get your watched repos fails, there's at least a button to refresh it. Maybe it should say something like "Error - Refresh"? Idk, there's such limited space here.

File: Classes/Notifications/NotificationSubscriptionsController.swift (hunk 6 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationSubscriptionsController.swift]: Ya, a `struct` would be better. Getting a little out of hand here.

File: Classes/Notifications/NotificationsViewController.swift (hunk 8 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController.swift]: Yup. In the repo notifications view, the left button will be "back". So I'm moving the "mark all" to the right so that it's consistent between views.

File: Classes/Notifications/NotificationsViewController.swift (hunk 10 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController.swift]: TIL!

File: Classes/Notifications/NotificationsViewController.swift (hunk 11 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController.swift]: Tbh I'm not sure, lol. I'm going to change it back and be "View Archived"

File: Classes/Notifications/NotificationsViewController.swift (hunk 13 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController.swift]: oops

File: Classes/Notifications/NotificationsViewController.swift (hunk 16 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController.swift]: Don't need to recalc every time in the loop. Grab it once locally and use that instead.

File: Classes/Notifications/NotificationsViewController.swift (hunk 14 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController.swift]: It's paged to 50 mixed-repo events, so number here likely wont be accurate if you have more than 50 unread notifications.

File: Classes/Notifications/NotificationSubscriptionsController.swift (hunk 4 above)
Comment (BasThomas) [reply Classes/Notifications/NotificationSubscriptionsController.swift]: Using a `State` enum inside of a `Controller` :)

File: Classes/Notifications/NotificationsViewController.swift (hunk 14 above)
Comment (BasThomas) [reply Classes/Notifications/NotificationsViewController.swift]: We \*could\* add the number for < 50 then maybe? 🤔

File: Classes/Notifications/NotificationsViewController.swift (hunk 16 above)
Comment (BasThomas) [reply Classes/Notifications/NotificationsViewController.swift]: Ah yep, didn't realise it's computed. 👍
